### PR TITLE
Tweak: [Core] [NC] [A11Y] Nested Carousel Accessibility [ED-10852]

### PR DESCRIPTION
**Pull request by hein-obox**
## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [x] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

File: includes/widgets/image-carousel.php [hunk 1]
```diff
@@ -931,26 +951,43 @@ protected function render() {
...
 				<?php echo implode( '', $slides ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>
 			</div>
 			<?php if ( 1 < $slides_count ) : ?>
-				<?php if ( $show_dots ) : ?>
-					<div class="swiper-pagination"></div>
-				<?php endif; ?>
 				<?php if ( $show_arrows ) : ?>
 					<div class="elementor-swiper-button elementor-swiper-button-prev" role="button" tabindex="0">
 						<?php $this->render_swiper_button( 'previous' ); ?>
-						<span class="elementor-screen-only"><?php echo esc_html__( 'Previous image', 'elementor' ); ?></span>
```
Comment (hein-obox) [inline includes/widgets/image-carousel.php:947]: These screen readers notifications have been replaced by a standard functionality from the SwiperJS library.

File: includes/widgets/image-carousel.php [hunk 2]
```diff
@@ -930,27 +941,25 @@ protected function render() {
...
 		?>
 		<div <?php $this->print_render_attribute_string( 'carousel-wrapper' ); ?>>
 			<div <?php $this->print_render_attribute_string( 'carousel' ); ?>>
 				<?php // PHPCS - $slides contains the slides content, all the relevent content is escaped above. ?>
 				<?php echo implode( '', $slides ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>
 			</div>
 			<?php if ( 1 < $slides_count ) : ?>
-				<?php if ( $show_dots ) : ?>
-					<div class="swiper-pagination"></div>
-				<?php endif; ?>
```
Comment (hein-obox) [inline includes/widgets/image-carousel.php:943]: I have moved pagination to below the navigation arrows, so that the navigation arrows will be accessed before the pagination for users who use keyboard to navigate the page.

**Review by TzviRabinovitch — DISMISSED** [review]

File: includes/widgets/image-carousel.php (hunk 2 above)
Comment (Barcov) [reply includes/widgets/image-carousel.php:943]: @hein-obox  Love this change

**Review by Barcov — DISMISSED** [review]

**Review by Barcov — DISMISSED** [review]

File: assets/dev/js/frontend/handlers/base-carousel.js [hunk 3]
```diff
@@ -199,4 +269,66 @@ export default class CarouselHandlerBase extends SwiperHandlerBase {
...
+	a11ySetWidgetAriaDetails() {
+		const $widget = this.$element;
+
+		$widget.attr( 'aria-roledescription', 'carousel' );
+		$widget.attr( 'aria-label', elementorFrontend.config.i18n.a11yCarouselWrapperAriaLabel );
+	}
+
+	a11ySetPaginationTabindex() {
+		this.getPaginationBullets().forEach( ( bullet ) => {
+			if ( ! bullet.classList.contains( 'swiper-pagination-bullet-active' ) ) {
```
[the remark this answers is not in the document]
Comment (louiswol94) [reply assets/dev/js/frontend/handlers/base-carousel.js:288]: @hein-obox Question: can we get these swiper classes fro the swiper configs instead of hard coding it?
Try: `swiper.params.pagination.bulletActiveClass` for example.

File: assets/dev/js/frontend/handlers/base-carousel.js [hunk 4]
```diff
@@ -199,4 +269,66 @@ export default class CarouselHandlerBase extends SwiperHandlerBase {
...
+				slide.setAttribute( 'aria-hidden', true );
+				slide.setAttribute( 'inert', '' );
+			} else {
+				slide.removeAttribute( 'aria-hidden' );
+				slide.removeAttribute( 'inert' );
+			}
+		} );
+	}
+
+	handleElementHandlers() {}
```
Comment (louiswol94) [inline assets/dev/js/frontend/handlers/base-carousel.js:333]: @hein-obox Maybe just add a comment here that it is an empty method so that child classes can override this method.

File: assets/dev/js/frontend/handlers/base-carousel.js [hunk 5]
```diff
@@ -89,11 +99,14 @@ export default class CarouselHandlerBase extends SwiperHandlerBase {
...
+		if ( showPagination ) {
 			swiperOptions.pagination = {
-				el: '.swiper-pagination',
-				type: 'bullets',
+				el: `.elementor-element-${ this.getID() } .swiper-pagination`,
+				type: !! elementSettings.pagination ? elementSettings.pagination : 'bullets',
 				clickable: true,
+				renderBullet: ( index, classname ) => {
+					return `<span class="${ classname }" data-bullet-index="${ index }" aria-label="${ elementorFrontend.config.i18n.a11yCarouselPaginationBulletMessage } ${ index + 1 }"></span>`;
+				},
```
Comment (louiswol94) [inline assets/dev/js/frontend/handlers/base-carousel.js:109]: @hein-obox Why is this needed? We are rendering this via PHP, no?

**Review by louiswol94 — CHANGES_REQUESTED** [review]

File: assets/dev/js/frontend/handlers/base-carousel.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (hein-obox) [reply assets/dev/js/frontend/handlers/base-carousel.js:288]: Updated.
I like it!

I haven't added the 'bullet?', because it is in a forEach loop. If there is no bullet, this line of code won't be triggered.

File: assets/dev/js/frontend/handlers/base-carousel.js [hunk 6]
```diff
@@ -125,6 +156,45 @@ export default class CarouselHandlerBase extends SwiperHandlerBase {
...
+		this.elements.$paginationWrapper.on( 'keydown', '.swiper-pagination-bullet', this.onDirectionArrowKeydown.bind( this ) );
+		this.elements.$swiperContainer.on( 'keydown', '.swiper-slide', this.onDirectionArrowKeydown.bind( this ) );
+		this.$element.find( ':focusable' ).on( 'focus', this.onFocusDisableAutoplay.bind( this ) );
+	}
+
+	unbindEvents() {
+		this.elements.$swiperArrows.off();
+		this.elements.$paginationWrapper.off();
+		this.elements.$swiperContainer.off();
+		this.$element.find( ':focusable' ).off();
```
[the remark this answers is not in the document]
Comment (hein-obox) [reply assets/dev/js/frontend/handlers/base-carousel.js:176]: I am not sure.
I always add the unbindEvents, but I don't know how how it really works...

File: assets/dev/js/frontend/handlers/base-carousel.js [hunk 7]
```diff
@@ -125,6 +156,45 @@ export default class CarouselHandlerBase extends SwiperHandlerBase {
...
+	onDirectionArrowKeydown( event ) {
+		const isRTL = elementorFrontend.config.isRTL,
+			inlineDirectionArrows = [ 'ArrowLeft', 'ArrowRight' ],
+			currentKeydown = event.originalEvent.code,
+			isDirectionInlineKeydown = -1 !== inlineDirectionArrows.indexOf( currentKeydown ),
+			directionStart = isRTL ? 'ArrowRight' : 'ArrowLeft',
+			directionEnd = isRTL ? 'ArrowLeft' : 'ArrowRight';
+
+		if ( ! isDirectionInlineKeydown ) {
+			return true;
```
[the remark this answers is not in the document]
Comment (hein-obox) [reply assets/dev/js/frontend/handlers/base-carousel.js:188]: If I don't added 'true', then all other key events get disabled. When I use 'true', e.g. tab key stays active.

File: assets/dev/js/frontend/handlers/base-carousel.js [hunk 8]
```diff
@@ -199,4 +269,66 @@ export default class CarouselHandlerBase extends SwiperHandlerBase {
 
 		this.swiper.update();
 	}
+
+	getPaginationBullets( array = true ) {
```
[the remark this answers is not in the document]
Comment (hein-obox) [reply assets/dev/js/frontend/handlers/base-carousel.js:273]: Neat!

Updated.

File: assets/dev/js/frontend/handlers/base-carousel.js [hunk 9]
```diff
@@ -199,4 +269,66 @@ export default class CarouselHandlerBase extends SwiperHandlerBase {
...
+	a11ySetPaginationTabindex() {
+		this.getPaginationBullets().forEach( ( bullet ) => {
+			if ( ! bullet.classList.contains( 'swiper-pagination-bullet-active' ) ) {
+				bullet.removeAttribute( 'tabindex' );
+			}
+		} );
+
+		const isDirectionInlineArrowKey = 'ArrowLeft' === event?.code || 'ArrowRight' === event?.code;
+
+		if ( event?.target.classList.contains( 'swiper-pagination-bullet' ) && isDirectionInlineArrowKey ) {
```
[the remark this answers is not in the document]
Comment (hein-obox) [reply assets/dev/js/frontend/handlers/base-carousel.js:295]: Implemented.

File: assets/dev/js/frontend/handlers/base-carousel.js (hunk 4 above)
Comment (hein-obox) [reply assets/dev/js/frontend/handlers/base-carousel.js:333]: Updated.

File: assets/dev/js/frontend/handlers/base-carousel.js (hunk 5 above)
Comment (hein-obox) [reply assets/dev/js/frontend/handlers/base-carousel.js:109]: 1. The bullets are generated with JS by the SwiperJS library.
2. I need to customise the html a little bit.
3. This is standard function provided by the SwiperJS API.

File: assets/dev/js/frontend/handlers/base-carousel.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (louiswol94) [reply assets/dev/js/frontend/handlers/base-carousel.js:176]: @hein-obox My concern is just that in `bindEvents()` you are running this.swiper.autoplay.stop();
So should you not run `this.swiper.autoplay.start();` in `unbindEvents()`

**Review by louiswol94 — DISMISSED** [review]

**Review by louiswol94 — APPROVED** [review]
